[Elgin Lee] iP

ElginL opened the pull request:
# DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/ElginL/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (Coming soon)
- [ ] Reminders (Coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

ElginL committed:
add Duke's ability to greet user

ElginL committed:
add Duke's ability to echo text and exit

ElginL committed:
add Duke's ability to add items and list when required

ElginL committed:
add Duke's ability to mark and unmark things and list accurately

ElginL committed:
add Task class which represents a class

ElginL committed:
add usage of Task class in Duke

ElginL committed:
add ToDo class (subtype of Task)

ElginL committed:
add Deadline class (Subtype of Task)

ElginL committed:
refactored toString in Task and changed usage in Duke

ElginL committed:
refactored ToDo class toString method

ElginL committed:
refactored deadline class toString method

ElginL committed:
add Event class that is a subtype of Task

ElginL committed:
add documentation to ToDo class

ElginL committed:
change duke to have tasks from event, todo and deadline

ElginL committed:
edit Event toString and attributes encapsulated

ElginL committed:
edit deadline toString format

ElginL committed:
add automated text ui testing (with input and expected output txt)

ElginL committed:
add DukeException

ElginL committed:
add error handling for user input for mark and create

ElginL committed:
add delete functionality, refactored startService into multiple methods to add/list/delete/etc.

ElginL committed:
changed string representation of Event

ElginL committed:
delete total task created tracking

ElginL committed:
updated expected.txt to fit change in string representation for event.

ElginL committed:
modify DukeException getMessage method

ElginL committed:
update Duke to give better style and formatting

ElginL committed:
Update expected results to pass file compare test

ElginL committed:
Update Deadline, Event, Task, ToDo

New constructors to initialize task name and isDone together as it
helps in preloading of tasks from file

ElginL committed:
Update Duke

Add preloading of tasks from file functionality from a text file

ElginL committed:
Update Duke.java

Commands (add/delete/mark/etc.) now saves to duke.txt file

ElginL committed:
Add duke.txt

File that contains some tasks for preloading

ElginL committed:
Update Deadline, Event

Deadline and Event stores date as LocalDate instead of String,
Constructor throws DateTimeParseException on failure to cast into
LocalDate

ElginL committed:
Update Duke.java

Handles cases where user input date is invalid

ElginL committed:
Merge branch 'branch-Level-7'

ElginL committed:
Merge branch 'branch-Level-8'

Resolve conflict in Duke.java, update constructor in Deadline and Event

ElginL committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

ElginL committed:
Add checkstyle with gradle

ElginL committed:
Fix styling in project

Duke.java and DukeException.java style changed according to
gradle style check

ElginL committed:
Add storage class

Handles saving and loading of tasks to Duke, for better OOP
practice

ElginL committed:
Update Duke

Uses Storage object to handle saving and loading of tasks

ElginL committed:
Add TaskList class

Handles add/delete/mark operations for tasks and stores all tasks

ElginL committed:
Modify Duke and Storage

Support the use of TaskList

ElginL committed:
Add Command class

ElginL committed:
Add AddCommand, DeleteCommand, ExitCommand, ListCommand, MarkCommand,
UnmarkCommand

Helps to execute commands to create an outcome

ElginL committed:
Add Parser class

Helps to makes sense of a user input

ElginL committed:
Add Ui class

Helps to handle interaction between Duke and User

ElginL committed:
Modify Deadline, Event, Task, ToDo

Add documentation for methods

ElginL committed:
Update DukeException

Removed OOPS!!! from getMessage because it is now added in Ui class

ElginL committed:
Update TaskList

Add documentation and modified some methods to return task

ElginL committed:
Update Duke

Incorporated all new Command classes, Ui, and Parser to serve
the main method

ElginL committed:
Refactor all classes into Duke package

ElginL committed:
Refactored classes into more packages

Now has command, exception, parser, storage, task and ui package

ElginL committed:
Add JUnit test for AddCommand and Parser

ElginL committed:
Add configurations to create JAR files

ElginL committed:
Modify AddCommandTest and ParserTest

Add JavaDoc to testing methods

ElginL committed:
Add Find Command

Command that helps to find matching descriptors in Task

ElginL committed:
Modify Parser

Parser now detects find command and returns it if called

ElginL committed:
Modify TaskList and Ui

TaskList now has find method and Ui now has a method that print
search results

ElginL committed:
Update duke.txt with new tasks

ElginL committed:
Merge branch 'branch-A-JavaDoc'

ElginL committed:
Merge branch 'branch-Level-9'

ElginL committed:
Modify all files to comply with coding standard

ElginL committed:
Merge branch 'branch-A-CodingStandard'

ElginL committed:
Update Ui

Change Javadoc comment for bye method

ElginL committed:
Merge branch 'A-CheckStyle'

prit3010 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public class AddCommand extends Command {
+    /** The name of the command (e.g. todo, deadline, event). */
+    private final String command;
```
Constants should be in uppercase

prit3010 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,193 @@
+     * @return String representation of all tasks
+     */
+    public String allItems() {
+        StringBuilder sb = new StringBuilder();
```
Good use of StringBuilder! It is faster than adding the strings up

prit3010 commented on src/main/java/duke/task/ToDo.java:
```diff
@@ -0,0 +1,37 @@
+        super(taskName, isDone);
+    }
+
+    /**
+     * String representation of a ToDo.
+     *
+     * @return ToDo string representation.
+     */
```
Since you are overriding this method, you can explore using `@inheritDoc` which will help you inherit the JavaDocs from the parent method

prit3010 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,162 @@
+     *
+     * @param error Error message.
+     */
+    public void showDukeError(String error) {
+        System.out.println(Ui.formatText("☹ OOPS!!! " + error));
+    }
+
+    /**
+     * Shows date parsing error (Invalid date format).
+     *
+     */
+    public void showInvalidDateError() {
+        String errorMessage = "☹ OOPS!!! Your date format has to be in the form 'yyyy-mm-dd'";
+        System.out.println(Ui.formatText(errorMessage));
+    }
+
+    /**
+     * Shows String to Number cast error (because user did not input a number).
+     *
+     */
+    public void showNumberCastError() {
+        String errorMessage = "☹ OOPS!!! Please input a valid index (i.e. a number)";
+        System.out.println(Ui.formatText(errorMessage));
+    }
```
These methods seem very similar. Perhaps you can abstract it out to just the `showDukeError`

prit3010 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,141 @@
+                String[] details = line.split(" \\| ");
+
+                try {
+                    switch (details[0]) {
+                    case "T":
+                        tempList.add(new ToDo(details[2], details[1].equals("1")));
+                        break;
+                    case "D":
+                        tempList.add(new Deadline(details[2], details[3], details[1].equals("1")));
+                        break;
+                    case "E":
+                        tempList.add(new Event(details[2], details[3], details[1].equals("1")));
+                        break;
+                    default:
+                        throw new DukeException("File contains lines that cannot be validated as a Task.");
+                    }
+                } catch (DukeException e) {
```
Good use of `switch`

prit3010 reviewed (commented):
Good Job! Your code adheres to the OOP principles very well with your command, exception, parser, storage and task files.  Code follows the coding standards very well too.

ElginL committed:
Add MainWindow.fxml, MainWindow.java, Main.java, Launcher.java

Shows a GUI with a TextInput and submit button, but does not show
the results yet

ElginL committed:
Update commands, Ui, fxml, controllers

Commands execute method now returns a string of what duke wants
to say
Ui now returns strings to texts instead of printing to command line
Controllers and fxml updated for cleaner ui

ElginL committed:
Update Javadoc

Update javadoc for all classes to pass checkstyleMain

ElginL committed:
Merge branch 'branch-Level-10'

ElginL committed:
Update Ui

Change method names to verbs and updated classes that use is affected
accordingly

linuschui commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,34 @@
+import javafx.fxml.FXMLLoader;
+import javafx.scene.Scene;
+import javafx.scene.layout.VBox;
+import javafx.stage.Stage;
```
Perhaps sort the import alphabetically? 😄 (for all other classes too!!)

linuschui commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,105 @@
+            "event", "deadline", "mark", "unmark", "delete", "find"));
+
+    /** All valid commands without arguments. */
+    private static final ArrayList<String> validCommandsWithoutArgs = new ArrayList<>(Arrays.asList("bye", "list"));
```
Perhaps variables declared as final should be capitalised? 😄

linuschui commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,118 @@
+     *
+     * @return Message conveying that user that has successfully deleted a task, and the number of tasks left
+     */
+    public String taskDeletedMsg(int newTaskCount, String taskDescription) {
```
Perhaps rename this to `getTaskDeletedMsg` to match the above method `getTaskAddedMsg` ? 😸

linuschui commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,118 @@
+     *
+     * @param taskDescription The description of the task unmarked.
+     */
+    public String getTaskUnmarkedMsg(String taskDescription) {
```
Perhaps rename Msg to Message so it will be clearer for people who might not be familiar with short forms? 😸

linuschui commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,193 @@
+     * @throws DukeException If index is not given, or index <= 1 or index >= tasks.size().
+     * @throws NumberFormatException If index given by user cannot be casted into an integer.
+     */
+    public String markOrUnmark(String userInput, boolean isMarkDone) throws DukeException, NumberFormatException {
```
perhaps change `isMarkDone` to `isMarked`? 😸

linuschui commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,118 @@
+     * @param error Error message.
+     * @return Formatted duke error message.
+     */
+    public String formatDukeErrorMsg(String error) {
```
Perhaps change `formatDukeErrorMsg` to `getFormatDukeErrorMsg` for consistency with other methods? 😄

ElginL committed:
Update Ui

Change some methods to have a verb prefix

ElginL committed:
Update build.gradle

Chane mainClassName to duke.Launcher

ElginL committed:
Add assertions

Currently, every class do not have any assertions

There is a need to add assertions because we need to make sure that
our assumptions about some code is correct and not faulty, to stop
it at development phase before publishing

Change
* AddCommand assert task created is not null
* DeleteCommand assert task deleted description is not empty string
* FindCommand assert results is not empty string
* MarkCommand assert task description is not empty string
* UnmarkCommand assert task description is not empty
* Storage assert else case is an event
* Ui assert that tasks listed is not empty

It is done this way in order to make sure that tasks are created,
marked, unmarked, deleted, saved and printed to the screen correctly

ElginL committed:
Update Deadline

Currently, the access modifier for date in Deadline is protected

It needs to be changed to private which is more secure than access
modifier. We do not want other classes to change this variable
directly.

Change protected access modifier to private

It is done this way to secure data within the class, and is the most
secure way to do it by the Java convention

ElginL committed:
Update Deadline, Event code quality

Add adherence to avoid complicated expressions

ElginL committed:
Update AddCommand and Storage code quality

Add compliance with using the default branch

ElginL committed:
Update Storage exception handling

Shift out handling of DukeException to be caught in Duke instead

ElginL committed:
Update use of class variables with static

Currently, classes that will only have one instance does not store
static fields

This needs to change because it is confusing for other developers
and it harnesses the use of static

Update classes that have only one instance to have static fields

This is done because it will make it clearer for other developers

ElginL committed:
Merge pull request #1 from ElginL/branch-A-Assertions

Branch A Assertions

ElginL committed:
Merge pull request #2 from ElginL/branch-A-CodeQuality

Branch A CodeQuality

ElginL committed:
Revert "Branch A CodeQuality"

ElginL committed:
Merge pull request #3 from ElginL/revert-2-branch-A-CodeQuality

Revert "Branch A CodeQuality"

ElginL committed:
Revert "Revert "Branch A CodeQuality""

ElginL committed:
Merge pull request #4 from ElginL/revert-3-revert-2-branch-A-CodeQuality

Revert "Revert "Branch A CodeQuality""

ElginL committed:
Update Storage and UnmarkCommand

Fix bugs causing compilation error

ElginL committed:
Add gradle.yml

Configuration for Github actions (continuous integration)

ElginL committed:
Add Loanbook, Contact, add to loanbook and list loanbook functionality

ElginL committed:
Add loanbook delete operation

ElginL committed:
Update Loanbook

Prevents duplicate names from being added to loanbook

ElginL committed:
Merge branch 'extension-B-Loanbook'

ElginL committed:
Add save functionality for loanbook, update parser and duke

ElginL committed:
Update Event and Deadline

Time of Event/Deadline present in tasks now

ElginL committed:
Update DialogBox, MainWindow

Improved UI of the application

ElginL committed:
Merge branch 'branch-A-BetterGui'

ElginL committed:
Update TaskList, Main

Add icon for application, and improved some toString styling

ElginL committed:
Update README.md

Add user guide for application

ElginL committed:
Update README.MD

Add missing user guide on bye command

ElginL committed:
Add HelpCommand

ElginL committed:
Update HelpCommand

Corrected link to documentation

ElginL committed:
Update Storage

Saves the file every time a command is run

ElginL committed:
Update Duke

Remove unused imports

ElginL committed:
Update docs/README.md

Add link in features for exit program command

ElginL committed:
Update README.md
